modules/gdscript/gdscript_parser.cpp

```diff
@@ -4342,6 +4342,13 @@ bool GDScriptParser::export_annotations(const AnnotationNode *p_annotation, Node
 					variable->export_info.type = Variant::OBJECT;
 					variable->export_info.hint = PROPERTY_HINT_RESOURCE_TYPE;
 					variable->export_info.hint_string = class_name;
+				} else if (class_name != StringName() && native_base != StringName() && ClassDB::is_parent_class(native_base, SNAME("Node"))) {
+					variable->export_info.type = Variant::OBJECT;
+					variable->export_info.hint = PROPERTY_HINT_NODE_TYPE;
+					variable->export_info.hint_string = class_name;
+				} else {
+					push_error(R"(Export type can only be built-in, a resource, a node, or an enum.)", p_annotation);
+					return false;
 				}
```

Comment:
Please add a check for global classes, which is currently missing. The inspector currently cannot work with non-global and inner classes.

```c++
				if (class_name != StringName()) {
					push_error(R"(Only global classes can be exported.)", p_annotation);
					return false;
				}
				if (ClassDB::is_parent_class(native_base, SNAME("Resource"))) {
					variable->export_info.type = Variant::OBJECT;
					variable->export_info.hint = PROPERTY_HINT_RESOURCE_TYPE;
					variable->export_info.hint_string = class_name;
				} else if (ClassDB::is_parent_class(native_base, SNAME("Node"))) {
					variable->export_info.type = Variant::OBJECT;
					variable->export_info.hint = PROPERTY_HINT_NODE_TYPE;
					variable->export_info.hint_string = class_name;
				} else {
					push_error(R"(Export type can only be built-in, a resource, a node, or an enum.)", p_annotation);
					return false;
				}
```

```c++
			case GDScriptParser::DataType::CLASS: {
				StringName class_name;
				if (export_type.class_type) {
					class_name = export_type.class_type->get_global_name();
				}
				if (class_name != StringName()) {
					push_error(R"(Only global classes can be exported.)", p_annotation);
					return false;
				}
				if (ClassDB::is_parent_class(export_type.native_type, SNAME("Resource"))) {
					variable->export_info.type = Variant::OBJECT;
					variable->export_info.hint = PROPERTY_HINT_RESOURCE_TYPE;
					variable->export_info.hint_string = class_name;
				} else if (ClassDB::is_parent_class(export_type.native_type, SNAME("Node"))) {
					variable->export_info.type = Variant::OBJECT;
					variable->export_info.hint = PROPERTY_HINT_NODE_TYPE;
					variable->export_info.hint_string = class_name;
				} else {
					push_error(R"(Export type can only be built-in, a resource, a node, or an enum.)", p_annotation);
					return false;
				}
			} break;
```

Also it would be nice to add tests, see [`export_variable.gd`](https://github.com/godotengine/godot/blob/master/modules/gdscript/tests/scripts/parser/features/export_variable.gd). To test `CLASS` you can add a `class_name` to the current script. To test `SCRIPT` you can add `.notest.gd` scripts and `const`-`preload` them.

Reply:
Makes sense, will get that set up asap, probably the weekend

Reply:
Sorry I totally blanked on the test part of this, unresolving so I can look at it again

Reply:
Finally got some time to put these tests in! Resolving the conversation, hope that's good to merge now :)